transport/httpcommon/diag.go

```diff
@@ -0,0 +1,198 @@
+			},
+			GotFirstResponseByte: func() {
+				logger.Print("Response started")
+			},
+			Got1xxResponse: func(code int, header textproto.MIMEHeader) error {
+				logger.Printf("Got info response status=%d, headers=%v", code, header)
+				return nil
+			},
+			DNSStart: func(info httptrace.DNSStartInfo) {
+				logger.Printf("Starting DNS lookup for %q", info.Host)
+			},
+			DNSDone: func(info httptrace.DNSDoneInfo) {
+				logger.Printf("Done DNS lookup: %+v", info)
+			},
+			ConnectStart: func(network, addr string) {
+				logger.Printf("Connection started to %s:%s", network, addr)
+			},
+			ConnectDone: func(network, addr string, err error) {
+				logger.Printf("Connection to %s:%s done, err: %v", network, addr, err)
+			},
+			TLSHandshakeStart: func() {
+				logger.Print("TLS handshake starting")
+			},
+			TLSHandshakeDone: func(state tls.ConnectionState, err error) {
+				logger.Printf("TLS handshake done. state=%+v err=%v", state, err)
+				logger.Printf("Peer certificate count %d", len(state.PeerCertificates))
+				for i, crt := range state.PeerCertificates {
+					logger.Printf("- Peer Certificate %d\n\t%s", i, tlscommon.CertDiagString(crt))
+				}
+
+				logger.Printf("Verified chains count: %d", len(state.VerifiedChains))
+				for i, chain := range state.VerifiedChains {
+					for j, crt := range chain {
+						logger.Printf("- Chain %d certificate %d\n\t%s", i, j, tlscommon.CertDiagString(crt))
+					}
+				}
+			},
+			WroteHeaders: func() {
+				logger.Printf("Wrote request headers")
+			},
+			Wait100Continue: func() {
+				logger.Printf("Waiting for continue")
+			},
+			WroteRequest: func(info httptrace.WroteRequestInfo) {
+				logger.Printf("Wrote request err=%v", info.Err)
+			},
+		}
+		for i, req := range reqs {
+			logger.Printf("Request %d to %s starting", i, req.URL.String())
+			req = req.WithContext(httptrace.WithClientTrace(req.Context(), ct))
+			if resp, err := rt.RoundTrip(req); err != nil {
+				logger.Printf("request %d error: %s", i, diagError(err))
+			} else if isGoHTTPResp(resp) {
+				resp.Body.Close()
+				logger.Printf("request %d error: HTTP request sent to HTTPS server.", i)
+			} else {
+				resp.Body.Close()
+				logger.Printf("request %d successful. status=%d", i, resp.StatusCode)
+			}
+		}
+		return b.Bytes()
+	}
+}
+
+// isGoHTTPResp detects if the response is one that a go http.Server sends if an HTTP request is made to an HTTPS server.
+// non Go servers may return a net.OpError instead.
+func isGoHTTPResp(r *http.Response) bool {
+	if r.StatusCode != http.StatusBadRequest {
+		return false
+	}
+	p, err := io.ReadAll(r.Body)
+	if err != nil {
+		return false
+	}
+	return string(p) == "Client sent an HTTP request to an HTTPS server.\n"
+}
+
+// diagError tries to diagnose the error and return a cause/possible cause in a human readable format.
+// If no matching errors are found err.Error is returned.
+func diagError(err error) string {
```

Comment:
@leehinman, I've tried to add this to explain errors based on the scenarios you provided (with test cases for most of them). we can add more in the future as well